### PR TITLE
X contents validation

### PR DESCRIPTION
**Pull request by bkmartinjr**
Proposed revision to test full contents of the X matrix in validation pass, plus presence contents.

Tested on a full census build, on the standard config, and it ran 5 hrs 10 minutes.

### CODE REVIEW

File: tools/cell_census_builder/validate.py [hunk 1]
```diff
@@ -276,59 +276,86 @@ def _validate_X_layers_contents_by_dataset(args: Tuple[str, str, Dataset, List[E
...
+                continue
+
+            X_raw = (
+                exp.ms[MEASUREMENT_RNA_NAME]
+                .X["raw"]
+                .read((obs_joinids.soma_joinid.to_numpy(), slice(None)))
+                .tables()
+                .concat()
+            )
+            data = X_raw["soma_data"].to_numpy()
```
Comment (Bento007) [inline tools/cell_census_builder/validate.py:307]: Isn't it expensive to expand out the whole sparse matrix?

File: tools/cell_census_builder/validate.py [hunk 2]
```diff
@@ -276,59 +276,86 @@ def _validate_X_layers_contents_by_dataset(args: Tuple[str, str, Dataset, List[E
...
+            # positionally re-index the vars/cols using the feature_id as the join key.
+            cols_by_position = (
+                expected_ad_var.join(all_var_ids.set_index("feature_id"))
+                .set_index("soma_joinid")
+                .index.get_indexer(cols)
+            )
+
+            # Assertion 1 - the contents of the X matrix are EQUAL for all var values
+            # present in the AnnData
+            assert (
```
Comment (Bento007) [inline tools/cell_census_builder/validate.py:333]: I [read](https://stackoverflow.com/questions/30685024/check-if-two-scipy-sparse-csr-matrix-are-equal) that both shape and nnz should be equal before doing this comparison. Shapes makes sense. However I don't see how asserting nnz help.

File: tools/cell_census_builder/validate.py [hunk 3]
```diff
@@ -276,59 +276,86 @@ def _validate_X_layers_contents_by_dataset(args: Tuple[str, str, Dataset, List[E
...
+                sparse.coo_matrix((data, (rows_by_position, cols_by_position)), shape=expected_ad_x.shape)
+                != expected_ad_x
+            ).nnz == 0, f"{eb.name}:{dataset.dataset_id} the X matrix elements are not equal."
+
+            # Assertion 2 - the contents of the X matrix are EMPTY for all var ids
+            # NOT present in the AnnData. Test by asserting that no col IDs contain
+            # a joinid not in the AnnData.
+            assert (
+                var_joinid_in_adata.all()
+                or not pd.Series(cols).isin(all_var_ids[~var_joinid_in_adata].soma_joinid).any()
```
Comment (Bento007) [inline tools/cell_census_builder/validate.py:343]: Can you explain this check for me and why we have an `or`?

**Review by Bento007 — COMMENTED** [review]
Looks good

File: tools/cell_census_builder/validate.py [hunk 4]
```diff
@@ -276,59 +276,86 @@ def _validate_X_layers_contents_by_dataset(args: Tuple[str, str, Dataset, List[E
...
-                expected_ad_x_coo = expected_ad_x.tocoo()
-                assert expected_ad_x_coo.nnz == actual_x_coo.nnz, error.format(
-                    "nnz", actual_x_coo.nnz, expected_ad_x.nnz
-                )
-                assert expected_ad_x_coo.shape == actual_x_coo.shape, error.format(
-                    "shape", actual_x_coo.shape, expected_ad_x_coo.shape
+                expected_ad_x, expected_ad_var = ad.raw.X, ad.raw.var
+
+            # get the joinids for the obs axis
+            obs_joinids = (
```
[the remark this answers is not in the document]
Comment (bkmartinjr) [reply tools/cell_census_builder/validate.py:288]: Yes -- it is simpler and has fewer dependencies. The query layer is a more complex class that wraps this same operation (plus some additional overhead).  It introduces additional dependencies into the builder.  I don't see any benefits.

File: tools/cell_census_builder/validate.py (hunk 1 above)
Comment (bkmartinjr) [reply tools/cell_census_builder/validate.py:307]: sure, but if you want to compare every value withe AnnData, you don't have any choice.  You must read it all in order to check for equality.

File: tools/cell_census_builder/validate.py (hunk 2 above)
Comment (bkmartinjr) [reply tools/cell_census_builder/validate.py:333]: You may be misunderstanding the code -- this an equality test implemented with `__ne__` and an nnz check.  This pattern prevents the equality check from having to materialize an full array of bools (remember, the equality check is per-element).

See for example, this [stack overflow answer](https://stackoverflow.com/questions/30685024/check-if-two-scipy-sparse-csr-matrix-are-equal).  

Highly suggest you study the sparse array implementation and documentation to get a better understanding of how it works.

File: tools/cell_census_builder/validate.py (hunk 3 above)
Comment (bkmartinjr) [reply tools/cell_census_builder/validate.py:343]: consider the case where _all_ of the values in the mask are True.  In that case, the second clause will fail (the arg to `isin` will be an empty set), so it is protected with the first clause.

**Review by Bento007 — APPROVED** [review]

**Pull request merged**